### PR TITLE
[Feature:InstructorUI] TA Beta Testing Date Constraints

### PR DESCRIPTION
**Pull request by Christian2147**
### Why is this Change Important & Necessary?
Partially addresses #12349
This PR specifically removes all constraints for the TA beta testing date for gradeables. This is important and necessary because instructors felt that the current UI for setting dates on gradeables was confusing, and they want to have more freedom over what dates they set for gradeables, even if they are invalid, since a lot of those dates may not matter for the specific class or assignment.

### What is the New Behavior?
Now, the TA beta testing date has no constraints, it can be set to whatever date regardless of other dates. When one of the lower dates is set to before the present day and the TA beta testing date is set to the future, the lower date wins when assigning a gradeable status on the gradeables page.

A date with actual constraints:
Constraint visible
Previously, the TA beta testing date also had a constraint (where its box would turn red and changes would fail to save) if it was after any of the dates below it (For example: open submission date).

Now, when the TA beta testing date is ahead of one of the dates below it, such as the open submission date (Only one that really matters now since all the other dates have constraints), it will display a message like the one shown here:
image

Additionally, when the TA beta testing date is in the future, and the open submission date is in the past, the open submission date wins and the assignment shows as open. This is also true for all dates below the TA viewing date.

### What steps should a reviewer take to reproduce or test the bug or new feature?

Log in as instructor. Go to the sample course, try editing all gradeables and try changing their TA viewing date in all different combinations. Make sure that not only can the date be changed to any valid date, but also that the listing of the gradeable on the gradeables page is valid. For example, if open submission date is in the past, and TA viewing date in the future, then the assignment should show under OPEN, not under FUTURE, and not under BETA. 

Secondly, On the gradeable page, make sure all the buttons are visible and function properly. For example, if the assignment shows as OPEN, all appropriate buttons should be visible and should function like normal.

Finally, go to a gradeable labeled FUTURE and a gradeable labeled BETA and try to copy the link. Then log out and log in as a student and try to paste the link. You should NOT be able to access the gradeable. If you can, please report it immediately.

Note: While testing, the only dates that should have full free access to be changed without causing any constraint issues currently are the grade inquiry dates and the ta view date. The rest of the dates still have constraints on them.

### Automated Testing & Documentation

This feature is sufficiently tested by unit tests and cypress tests. Cypress tests have been added to ensure that the gradables are visible to the proper users when the TA beta testing date is changed along with the other dates, especially in ways that would previously violate a constraint. I have also added additional cypress tests for lab and other gradeables to ensure that the TA view date along with the other dates are working properly without the constraint. These tests also mainly consist of moving the dates around and checking to ensure the gradeable is viewed by the proper access levels.

### Other information

This is not a breakage change. However, this PR DID add a create a new migration for the course database. It removed any DB constraints from the gradeable table relating to g_ta_view_start_date.
Link hijacking might be a security concern with this PR and should be tested. Ensure that ONLY instructors have access to FUTURE gradeables and instructors and TAs/Mentors have access to BETA gradeables.

### CODE REVIEW

**Review by williamschen23 — CHANGES_REQUESTED** [review]
DB Dump needs to be fixed

Image
Missing edge case for simple gradeable, banner always displays. `submission_open_date` doesnt exist for simple gradeables.

**Review by JManion32 — COMMENTED** [review]
I noticed when I set the TA Beta testing Date Constraint to the end of time on a gradeable in the "FUTURE" section and then click `OPEN TO TAS NOW`, the gradeable jumps to the bottom of the page in the "Grades Available" section. Is this intended functionality since it would compare all of the dates and land on "Grades Available" since it is supposed to be the last one in chronological order?

I'd also like to know more about automated testing. The gradeables page is critical to courses, and gradeable date constraints have been reported to cause crashes (see #12204 and #10874). Additionally, there is a lot of edge cases that would be hard to adequately test manually.

Have you checked to see what E2E tests we currently have for changing gradeable dates? There appears to already be some in `edit_gradeable.spec.js`. It is definitely worth expanding on whatever exists now so that you have a nice foundation for this PR and all follow-ups for other constraints.

**Comment by williamschen23** [comment]
Ive added some specifically for grade inquries when I first made the first remove constraints PR. These definietely need more E2e tests, but I haven't been looking at this since CI is failing

**Comment by Christian2147** [comment]
> I noticed when I set the TA Beta testing Date Constraint to the end of time on a gradeable in the "FUTURE" section and then click `OPEN TO TAS NOW`, the gradeable jumps to the bottom of the page in the "Grades Available" section. Is this intended functionality since it would compare all of the dates and land on "Grades Available" since it is supposed to be the last one in chronological order?
> 
> I'd also like to know more about automated testing. The gradeables page is critical to courses, and gradeable date constraints have been reported to cause crashes (see #12204 and #10874). Additionally, there is a lot of edge cases that would be hard to adequately test manually.
> 
> Have you checked to see what E2E tests we currently have for changing gradeable dates? There appears to already be some in `edit_gradeable.spec.js`. It is definitely worth expanding on whatever exists now so that you have a nice foundation for this PR and all follow-ups for other constraints.

After some testing turns out this button just does this regardless of what you set the dates to. I will look into this today.

**Comment by Christian2147** [comment]
> Ive added some specifically for grade inquries when I first made the first remove constraints PR. These definietely need more E2e tests, but I haven't been looking at this since CI is failing

All CIs should be fixed now.

File: site/app/models/gradeable/Gradeable.php [hunk 1]
```diff
@@ -2389,7 +2366,13 @@ public function canView(User $user): bool {
 
         //If we're not instructor and this is not open to TAs
         $date = $this->core->getDateTimeNow();
-        if ($this->getTaViewStartDate() > $date && !$user->accessAdmin()) {
+        if ($this->getTaViewStartDate() > $date && $this->getSubmissionOpenDate() > $date && !$user->accessAdmin()) {
+            return false;
+        }
+
+        // If the gradeable has NO open submission date and TA view start date is in the future
+        $date = $this->core->getDateTimeNow();
```
Comment (williamschen23) [inline site/app/models/gradeable/Gradeable.php]: redeclaration of date is not necessary

File: site/app/models/gradeable/Gradeable.php [hunk 2]
```diff
@@ -2389,7 +2366,13 @@ public function canView(User $user): bool {
...
+        if ($this->getType() !== GradeableType::ELECTRONIC_FILE && $this->getTaViewStartDate() > $date && !$user->accessAdmin()) {
```
Comment (williamschen23) [inline site/app/models/gradeable/Gradeable.php]: this lowkey could be shortened

File: site/cypress/e2e/Cypress-Gradeable/edit_gradeable.spec.js [hunk 3]
```diff
@@ -243,8 +243,17 @@ describe('Tests cases revolving around modifying gradeables', () => {
...
+        updateDates('#date_ta_view', future_date, 'All Changes Saved');
+
+        // The gradeable should be visible to everyone (Submission open date get priority in listing over ta view date)
+        ['student', 'grader', 'ta'].forEach((user) => {
+            logoutLogin(user, ['sample']);
+            cy.get('#gradeables-content').should('contain.text', 'Open Peer Homework');
+        });
+
+        logoutLogin('instructor', ['sample', 'gradeable', 'open_peer_homework', 'update?nav_tab=5']);
+
```
Comment (williamschen23) [inline site/cypress/e2e/Cypress-Gradeable/edit_gradeable.spec.js:256]: we dont test if students can see gradeable that has not been open for TAs yet?

**Review by williamschen23 — CHANGES_REQUESTED** [review]

File: site/cypress/e2e/Cypress-Gradeable/edit_gradeable.spec.js (hunk 3 above)
Comment (Christian2147) [reply site/cypress/e2e/Cypress-Gradeable/edit_gradeable.spec.js:256]: I went over with Cutler about this. If the submission open date is in the past and the TA view date is in the future (Something that previously was not allowed), then the submission open date takes priority when determining if a gradeable can be viewed. That is what I am testing here.

File: migration/migrator/migrations/course/20260129214754_drop_ta_view_start_constraint.py [hunk 4]
```diff
@@ -0,0 +1,33 @@
...
+    :param config: Object holding configuration details about Submitty
+    :type config: migrator.config.Config
+    :param database: Object for interacting with given database for environment
+    :type database: migrator.db.Database
+    :param semester: Semester of the course being migrated
+    :type semester: str
+    :param course: Code of course being migrated
+    :type course: str
+    """
+    database.execute("ALTER TABLE gradeable ADD CONSTRAINT g_ta_view_start_date CHECK (g_ta_view_start_date <= g_grade_start_date);")
```
Comment (JManion32) [inline migration/migrator/migrations/course/20260129214754_drop_ta_view_start_constraint.py:33]: We discussed the down migration several weeks ago during our meeting. Barb brought up the risk of having to rollback, causing courses with abnormal dates to crash, but I think we landed on leaving it as-is right? Just double checking.

Comment (Christian2147) [reply migration/migrator/migrations/course/20260129214754_drop_ta_view_start_constraint.py:33]: I can clarify with her today. Either we leave it as is or we just make it do nothing.

**Comment by Christian2147** [comment]
> The code looks good to me.
> 
> I ran the Cypress test twice locally. It worked the first time (though an unrelated test failed). The second time, the first two passed, and the rest, including the new one, failed. Not sure if this is what the results have been for you. I am on a fresh VM snapshot.
> 
> We need more functionality reviews here, and it would be nice to have that Cypress test working reliably locally, but that may not be feasible if other tests aren't being atomic.

I am honestly having trouble getting any of the cypress tests anywhere to pass consistently locally. But I know that on github, the edit-gradeable.spec tests are consistently passing, and I have ran them at least twice I believe.

**Review by aconfo — APPROVED** [review]
See earlier comment

**Review by williamjallen — APPROVED** [review]
Code looks good to me.  As noted previously, the down migration is potentially risky.  I'll leave the decision about whether that risk is acceptable to Barb.
